### PR TITLE
Feature/blog setup

### PR DESCRIPTION
**Pull request by paulbrighton**

### CODE REVIEW

File: app/assets/javascripts/admin/users.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (krisquigley) [inline app/assets/javascripts/admin/users.coffee]: remove this file, in future avoid using controller generators

File: app/assets/javascripts/articles.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (krisquigley) [inline app/assets/javascripts/articles.coffee]: remove this file, in future avoid using controller generators

File: app/assets/javascripts/comments.coffee [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (krisquigley) [inline app/assets/javascripts/comments.coffee]: remove this file, in future avoid using controller generators

File: app/assets/stylesheets/admin/application.scss [hunk 4]
```diff
@@ -0,0 +1,61 @@
+ // Custom bootstrap variables must be set or imported *before* bootstrap.
+ @import "bootstrap";
+
+/* Admin page styling*/
```
Comment (krisquigley) [inline app/assets/stylesheets/admin/application.scss]: move this into it's own scss file

File: app/assets/stylesheets/admin/users.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the admin/users controller here.
```
Comment (krisquigley) [inline app/assets/stylesheets/admin/users.scss]: remove this file, in future avoid using controller generators

File: app/assets/stylesheets/application.scss [hunk 6]
```diff
@@ -10,6 +10,7 @@
  * files in this directory. Styles in this file should be added after the last require_* statement.
  * It is generally better to create a new file per style scope.
  *
- *= require_tree .
- *= require_self
  */
+
```
Comment (krisquigley) [inline app/assets/stylesheets/application.scss]: The above text can all be removed

File: app/assets/stylesheets/articles.scss [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Articles controller here.
```
Comment (krisquigley) [inline app/assets/stylesheets/articles.scss]: remove this file, in future avoid using controller generators

File: app/assets/stylesheets/comments.scss [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the comments controller here.
```
Comment (krisquigley) [inline app/assets/stylesheets/comments.scss]: remove this file, in future avoid using controller generators

File: app/controllers/admin/application_controller.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class Admin::ApplicationController < ActionController::Base
+  http_basic_authenticate_with name: "user", password: "password"
```
Comment (krisquigley) [inline app/controllers/admin/application_controller.rb]: Look at using environment variables for the name and password values

File: app/controllers/admin/articles_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      flash[:notice] = "You successfully created a post."
+      redirect_to @article
+    else
+      flash[:notice] = "Error, please try again."
+      render :new
+    end
+  end
+
+  def show
+    @article = Article.find(params[:id])
```
Comment (krisquigley) [inline app/controllers/admin/articles_controller.rb]: Look at moving this into a before action for show, edit, update and destroy

File: app/controllers/admin/articles_controller.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    if @article.save(article_params)
+      flash[:notice] = "You successfully created a post."
+      redirect_to @article
+    else
+      flash[:notice] = "Error, please try again."
+      render :new
+    end
+  end
+
+  def show
```
Comment (krisquigley) [inline app/controllers/admin/articles_controller.rb]: I would remove this action and instead have the admin panel link to the articles controller show instead

File: app/helpers/admin/users_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module Admin::UsersHelper
```
Comment (krisquigley) [inline app/helpers/admin/users_helper.rb]: remove this file, in future avoid using controller generators

File: app/helpers/articles_helper.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+module ArticlesHelper
```
Comment (krisquigley) [inline app/helpers/articles_helper.rb]: remove this file, in future avoid using controller generators

File: app/helpers/comments_helper.rb [hunk 14]
```diff
@@ -0,0 +1,2 @@
+module CommentsHelper
```
Comment (krisquigley) [inline app/helpers/comments_helper.rb]: remove this file, in future avoid using controller generators

File: app/views/articles/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,24 @@
+<h1>List of all Articles</h1>
+
+<table>
+  <thead>
+    <tr>
+      <th>Title</th>
+      <th>Author</th>
+      <!--<th>Body</th>-->
```
Comment (krisquigley) [inline app/views/articles/index.html.erb]: Remove this rather than commenting it out

File: app/views/comments/_form.html.erb [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+<%= form_for [ @article, @comment ] do |f| %>
+
+  <p>
+    <%= f.label :author_name %><br/>
+    <%= f.text_field :author_name %>
+  </p>
+
+  <p>
+    <%= f.label :body %><br/>
+    <%= f.text_area :body %>
```
Comment (krisquigley) [inline app/views/comments/_form.html.erb]: Add HTML 5 validations

File: app/views/admin/articles/_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,22 @@
+<%= form_for [:admin, @article] do |f| %>
+  <%= render partial: "shared/form_errors", object: @article %>
+  
+  <p>
+    <%= f.label :title %><br>
+    <%= f.text_field :title %>
```
Comment (krisquigley) [inline app/views/admin/articles/_form.html.erb]: Add HTML 5 validations

File: app/views/layouts/admin/application.html.erb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    <%= javascript_include_tag 'admin/application', 'data-turbolinks-track': 'reload' %>
+  </head>
+
+  <body>
+    <header>
+      <nav class="navbar navbar-light " style="background-color: #F26144;">
+        <div class="container-fluid">
+          <h1 class="admin__h1">Paul's Blog Admin Panel</h1>
+          <ul>
+            <li><a href="#">Articles</a></li>
```
Comment (krisquigley) [inline app/views/layouts/admin/application.html.erb]: use <%= link_to

File: config/environments/development.rb [hunk 19]
```diff
@@ -57,5 +57,5 @@
 
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
-  config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+  # config.file_watcher = ActiveSupport::EventedFileUpdateChecker
```
Comment (krisquigley) [inline config/environments/development.rb]: You can probably reeneable this again now

File: spec/features/admin/admin_articles_spec.rb [hunk 20]
```diff
@@ -0,0 +1,91 @@
+require 'rails_helper'
+include AuthHelper
+
+RSpec.describe "Articles", type: :feature do
+  before(:each) do
+    visit admin_articles_path
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: This probably isn't needed

File: spec/features/admin/admin_articles_spec.rb [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+      end
+    end
+  end
+  
+  describe "viewing an article" do
+    # create article
+    let!(:article) { create(:article) }
+
+    it "should show the contents of the article" do
+      http_login
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: This is already in the before each above

File: spec/features/admin/admin_articles_spec.rb [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+      expect(page).to have_content(article.text)
+      expect(page).to have_content(article.author)
+    end
+  end
+
+  describe "editing an article" do
+    let!(:article) { create(:article) }
+
+    it "should update the contents of an article" do
+      http_login
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: This is already in the before each above

File: spec/features/admin/admin_articles_spec.rb [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+      click_on "Update Article"
+
+      expect(page).to have_content("New Author")
+    end
+  end
+
+  describe "invalid data in article" do
+
+    it "should not create an article with invalid inputs" do
+      http_login
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: This is already in the before each above

File: spec/features/admin/admin_articles_spec.rb [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+      expect(page).to have_content("Author can't be blank")
+      expect(page).to have_content("Text can't be blank")
+    end
+  end
+
+  describe "deleting an article", js: true do
+    let!(:article) { create(:article) }
+
+    it "should delete the article" do
+      http_login
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: This is already in the before each above

File: spec/spec_helper.rb [hunk 25]
```diff
@@ -0,0 +1,101 @@
+require 'capybara/rspec'
```
Comment (krisquigley) [inline spec/spec_helper.rb]: These 2 lines should be in your capybara.rb file

File: test/controllers/admin/users_controller_test.rb [hunk 26]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class Admin::UsersControllerTest < ActionDispatch::IntegrationTest
```
Comment (krisquigley) [inline test/controllers/admin/users_controller_test.rb]: remove this file, in future avoid using controller generators

File: test/controllers/articles_controller_test.rb [hunk 27]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ArticlesControllerTest < ActionDispatch::IntegrationTest
```
Comment (krisquigley) [inline test/controllers/articles_controller_test.rb]: remove this file, in future avoid using controller generators

File: test/controllers/comments_controller_test.rb [hunk 28]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class CommentsControllerTest < ActionDispatch::IntegrationTest
```
Comment (krisquigley) [inline test/controllers/comments_controller_test.rb]: remove this file, in future avoid using controller generators

File: test/fixtures/articles.yml [hunk 29]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
```
Comment (krisquigley) [inline test/fixtures/articles.yml]: remove this file, in future avoid using controller generators

File: test/fixtures/comments.yml [hunk 30]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
```
Comment (krisquigley) [inline test/fixtures/comments.yml]: THe whole test folder can be deleted

File: Gemfile [hunk 31]
```diff
@@ -3,6 +3,8 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
 ruby '2.3.4'
 
+gem 'dotenv-rails', groups: [:development, :test]
```
Comment (krisquigley) [inline Gemfile]: Cleaner to move this into the `development, test` block below

File: app/assets/stylesheets/admin/application.scss [hunk 32]
```diff
@@ -1,61 +1,2 @@
...
-   text-align: center;
-   padding-top: 20px;
- }
-
- .new__post--link:hover {
-   color: #000;
- }
\ No newline at end of file
+// Custom bootstrap variables must be set or imported *before* bootstrap.
+@import "bootstrap";
```
Comment (krisquigley) [inline app/assets/stylesheets/admin/application.scss]: You'll need to import your admin.scss file here too

File: app/controllers/admin/articles_controller.rb [hunk 33]
```diff
@@ -9,29 +10,21 @@ def create
...
 
   def edit
-    @article = Article.find(params[:id])
   end
 
   def update
-    @article = Article.find(params[:id])
     
     if @article.update(article_params)
       flash[:notice] = "You successfully updated a post."
```
Comment (krisquigley) [inline app/controllers/admin/articles_controller.rb]: Follow the same pattern in all your controllers, the one below in the destroy action would be a good one to use everywhere

File: app/views/admin/articles/_form.html.erb [hunk 34]
```diff
@@ -2,21 +2,21 @@
   <%= render partial: "shared/form_errors", object: @article %>
   
   <p>
-    <%= f.label :title %><br>
-    <%= f.text_field :title %>
+    <%= f.label :title, class: "form__label"%><br>
```
Comment (krisquigley) [inline app/views/admin/articles/_form.html.erb]: You won't need this class if you use form groups from bootstrap

File: app/views/admin/articles/index.html.erb [hunk 35]
```diff
@@ -26,7 +26,6 @@
       </table>
     </div>
  
-
   <%= link_to 'New Post', new_admin_article_path, class: "new__post--link" %>
```
Comment (krisquigley) [inline app/views/admin/articles/index.html.erb]: I would use the bootstrap button classes for this

File: app/views/layouts/admin/application.html.erb [hunk 36]
```diff
@@ -6,16 +6,26 @@
     <%= csp_meta_tag %>
     <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
     <%= stylesheet_link_tag    'admin/application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag    'admin/admin', media: 'all', 'data-turbolinks-track': 'reload' %>
```
Comment (krisquigley) [inline app/views/layouts/admin/application.html.erb]: if you import this in `application.scss` then you wont need this link tag

File: config/initializers/assets.rb [hunk 37]
```diff
@@ -9,6 +9,7 @@
 Rails.application.config.assets.paths << Rails.root.join('node_modules')
 
 # Precompile additional assets.
-# application.js, application.css, and all non-JS/CSS in the app/assets
+# application.js, application.css and all non-JS/CSS in the app/assets
 # folder are already added.
 # Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += %w( admin/admin.css )
```
Comment (krisquigley) [inline config/initializers/assets.rb]: This can be removed after you've added the file to application.scss

File: spec/examples.txt [hunk 38]
```diff
@@ -1,8 +1,8 @@
 example_id                                            | status | run_time        |
 ----------------------------------------------------- | ------ | --------------- |
-./spec/features/admin/admin_articles_spec.rb[1:1:1]   | passed | 0.00827 seconds |
-./spec/features/admin/admin_articles_spec.rb[1:2:1:1] | passed | 0.42617 seconds |
-./spec/features/admin/admin_articles_spec.rb[1:3:1]   | passed | 0.01704 seconds |
-./spec/features/admin/admin_articles_spec.rb[1:4:1]   | passed | 0.04112 seconds |
-./spec/features/admin/admin_articles_spec.rb[1:5:1]   | passed | 0.02246 seconds |
-./spec/features/admin/admin_articles_spec.rb[1:6:1]   | passed | 1.54 seconds    |
+./spec/features/admin/admin_articles_spec.rb[1:1:1]   | passed | 0.30432 seconds |
```
Comment (krisquigley) [inline spec/examples.txt]: I would add this file to your gitignore

File: spec/features/admin/admin_articles_spec.rb [hunk 39]
```diff
@@ -3,15 +3,14 @@
...
-    visit admin_articles_path
     http_login
   end
 
   describe "listing articles" do
     it "should show the articles" do
       visit admin_articles_path
       
-      expect(page).to have_css "h2", text: "List of all Articles"
+      expect(page).to have_css "h2", text: "Articles"
```
Comment (krisquigley) [inline spec/features/admin/admin_articles_spec.rb]: I would check for an article title in this list instead

File: spec/support/auth_helper.rb [hunk 40]
```diff
@@ -1,7 +1,8 @@
 module AuthHelper
   def http_login
-    username = 'user'
-    password = 'password'
+    username = ENV['ADMIN_NAME']
```
Comment (krisquigley) [inline spec/support/auth_helper.rb]: Use fetch here so that the test suite fails if the environment variables are missing

File: Gemfile (hunk 31 above)
Comment (paulbrighton) [reply Gemfile]: I actually put it there first but the GitHub page instruction was to put it at the top.

File: app/views/comments/_form.html.erb [hunk 41]
```diff
@@ -2,17 +2,17 @@
 <%= render partial: "shared/form_errors", object: @comment %>
 <%= form_for [ @article, @comment ] do |f| %>
 
-  <p>
+  <div>
     <%= f.label :author_name %><br/>
-    <%= f.text_field :author_name %>
-  </p>
+    <%= f.text_field :author_name, required: :true, placeholder: :"Add your name here" %>
```
Comment (krisquigley) [inline app/views/comments/_form.html.erb]: For the placeholder text, just a string is fine:

`placeholder: "Add your name here"`

**Pull request merged**